File: employeeImportanceBFS.java
//very simple direct bfs solution where we use hashmap to store all employee objects with id as key.
// because we get id as arguement we need to get importance based on id we get
// note that subordinates are only direct subordinates

//sc and tc: o(n)

class Solution {
    public int getImportance(List<Employee> employees, int id) {
        if(employees == null || employees.size() == 0) {
            return 0;
        }
        HashMap<Integer, Employee> map = new HashMap<>();
        for(Employee e : employees) {
            map.put(e.id, e);
        }
        
        int total = 0;
        Queue<Integer> q = new LinkedList<>();
        q.add(id);
        
        while(!q.isEmpty()) {
            
            int cur = q.poll();
            Employee emp = map.get(cur);
            total += emp.importance;
            
            for(int juniors : emp.subordinates) {
                q.add(juniors);
            }
        }
        return total;
    }
}

File: employeeImportanceBFS.py
 """
# Definition for Employee.
class Employee:
    def __init__(self, id: int, importance: int, subordinates: List[int]):
        self.id = id
        self.importance = importance
        self.subordinates = subordinates
"""

class Solution:
    def getImportance(self, employees: List['Employee'], id: int) -> int:
        
        if not employees:
            return None
        result = 0
        q = list()
        dataStore = dict()
        
        for e in employees:
            dataStore[e.id] = e

        q.append(id)
        
        while q:
            curr = q.pop(0)
            e = dataStore.get(curr)
            result += e.importance
            for value in e.subordinates:
                q.append(value)
        
        return result
            
#      BFS approach, time complexity is O(n) and space complexity is O(n)
        
        
